File: ios/Plugin/ProxyManager.swift
import Foundation
import Capacitor

/**
 * Proxy Manager for handling network proxy configurations on iOS
 * Supports HTTP, HTTPS, SOCKS4, and SOCKS5 proxies
 */
public class ProxyManager {
    
    // MARK: - Properties
    
    private var enabled: Bool = false
    private var type: String = "http"
    private var host: String = ""
    private var port: Int = 8080
    private var username: String?
    private var password: String?
    private var bypass: [String] = []
    private var applyToAll: Bool = true
    
    // Statistics
    private var requestCount: Int = 0
    private var lastSuccessTime: Date?
    private var lastError: String?
    
    // MARK: - Initialization
    
    public init() {}
    
    // MARK: - Configuration
    
    /**
     * Configure global proxy settings
     */
    public func setConfig(_ config: JSObject?) {
        guard let config = config else {
            clearConfig()
            return
        }
        
        enabled = config["enabled"] as? Bool ?? false
        type = config["type"] as? String ?? "http"
        host = config["host"] as? String ?? ""
        port = config["port"] as? Int ?? 8080
        username = config["username"] as? String
        password = config["password"] as? String
        applyToAll = config["applyToAll"] as? Bool ?? true
        
        if let bypassArray = config["bypass"] as? [String] {
            bypass = bypassArray
        } else {
            bypass = []
        }
        
        print("[ProxyManager] Proxy configured: \(type)://\(host):\(port) (enabled: \(enabled))")
    }
    
    /**
     * Clear proxy configuration
     */
    public func clearConfig() {
        enabled = false
        type = "http"
        host = ""
        port = 8080
        username = nil
        password = nil
        bypass = []
        applyToAll = true
        lastError = nil
        
        print("[ProxyManager] Proxy configuration cleared")
    }
    
    /**
     * Get current proxy configuration as JSObject
     */
    public func getConfig() -> JSObject? {
        guard enabled else { return nil }
        
        var config: JSObject = [
            "enabled": enabled,
            "type": type,
            "host": host,
            "port": port,
            "applyToAll": applyToAll,
            "bypass": bypass
        ]
        
        if let username = username {
            config["username"] = username
        }
        
        return config
    }
    
    /**
     * Get proxy status
     */
    public func getStatus() -> JSObject {
        var status: JSObject = [
            "active": enabled && !host.isEmpty,
            "requestCount": requestCount
        ]
        
        if let lastError = lastError {
            status["lastError"] = lastError
        }
        
        if let lastSuccessTime = lastSuccessTime {
            status["lastSuccessTime"] = lastSuccessTime.timeIntervalSince1970 * 1000
        }
        
        if enabled, let config = getConfig() {
            status["config"] = config
        }
        
        return status
    }
    
    /**
     * Check if proxy is enabled and configured
     */
    public var isEnabled: Bool {
        return enabled && !host.isEmpty
    }
    
    // MARK: - Bypass Logic
    
    /**
     * Check if URL should bypass proxy
     */
    public func shouldBypass(url: String) -> Bool {
        guard !bypass.isEmpty else { return false }
        
        guard let parsedUrl = URL(string: url), let urlHost = parsedUrl.host else {
            return false
        }
        
        for pattern in bypass {
            if matchesPattern(host: urlHost, pattern: pattern) {
                return true
            }
        }
        
        return false
    }
    
    /**
     * Simple wildcard pattern matching for bypass list
     */
    private func matchesPattern(host: String, pattern: String) -> Bool {
        if pattern.hasPrefix("*.") {
            // Wildcard domain match
            let suffix = String(pattern.dropFirst(1))
            return host.hasSuffix(suffix) || host == String(pattern.dropFirst(2))
        } else if pattern.hasSuffix(".*") {
            // Wildcard IP match
            let prefix = String(pattern.dropLast(1))
            return host.hasPrefix(prefix)
        } else {
            return host.lowercased() == pattern.lowercased()
        }
    }
    
    // MARK: - URLSession Configuration
    
    /**
     * Apply proxy configuration to URLSessionConfiguration
     */
    public func applyToSessionConfig(_ config: URLSessionConfiguration) {
        guard isEnabled else { return }
        
        var proxyDict: [AnyHashable: Any] = [:]
        
        switch type.lowercased() {
        case "socks4", "socks5", "socks":
            proxyDict[kCFStreamPropertySOCKSProxyHost as String] = host
            proxyDict[kCFStreamPropertySOCKSProxyPort as String] = port
            
            if let username = username, !username.isEmpty {
                proxyDict[kCFStreamPropertySOCKSUser as String] = username
                proxyDict[kCFStreamPropertySOCKSPassword as String] = password ?? ""
            }
            
        case "https":
            proxyDict[kCFNetworkProxiesHTTPSEnable as String] = true
            proxyDict[kCFNetworkProxiesHTTPSProxy as String] = host
            proxyDict[kCFNetworkProxiesHTTPSPort as String] = port
            
        case "http":
            fallthrough
        default:
            proxyDict[kCFNetworkProxiesHTTPEnable as String] = true
            proxyDict[kCFNetworkProxiesHTTPProxy as String] = host
            proxyDict[kCFNetworkProxiesHTTPPort as String] = port
        }
        
        config.connectionProxyDictionary = proxyDict
        
        print("[ProxyManager] Proxy applied to session config: \(type)://\(host):\(port)")
    }
    
    /**
     * Apply proxy configuration considering per-request config and bypass list
     */
    public func applyToSessionConfig(_ config: URLSessionConfiguration, url: String, requestProxy: JSObject?) {
        // Check if request has its own proxy config
        if let requestProxy = requestProxy, requestProxy["enabled"] as? Bool == true {
            applyRequestProxy(config, proxyConfig: requestProxy)
            return
        }
        
        // Check if global proxy should apply
        guard isEnabled && applyToAll else { return }
        
        // Check bypass list
        if shouldBypass(url: url) {
            print("[ProxyManager] Bypassing proxy for URL: \(url)")
            return
        }
        
        applyToSessionConfig(config)
    }
    
    /**
     * Apply per-request proxy configuration
     */
    private func applyRequestProxy(_ config: URLSessionConfiguration, proxyConfig: JSObject) {
        let proxyType = proxyConfig["type"] as? String ?? "http"
        let proxyHost = proxyConfig["host"] as? String ?? ""
        let proxyPort = proxyConfig["port"] as? Int ?? 8080
        
        guard !proxyHost.isEmpty else { return }
        
        var proxyDict: [AnyHashable: Any] = [:]
        
        switch proxyType.lowercased() {
        case "socks4", "socks5", "socks":
            proxyDict[kCFStreamPropertySOCKSProxyHost as String] = proxyHost
            proxyDict[kCFStreamPropertySOCKSProxyPort as String] = proxyPort
            
            if let username = proxyConfig["username"] as? String, !username.isEmpty {
                proxyDict[kCFStreamPropertySOCKSUser as String] = username
                proxyDict[kCFStreamPropertySOCKSPassword as String] = proxyConfig["password"] as? String ?? ""
            }
            
        case "https":
            proxyDict[kCFNetworkProxiesHTTPSEnable as String] = true
            proxyDict[kCFNetworkProxiesHTTPSProxy as String] = proxyHost
            proxyDict[kCFNetworkProxiesHTTPSPort as String] = proxyPort
            
        default:
            proxyDict[kCFNetworkProxiesHTTPEnable as String] = true
            proxyDict[kCFNetworkProxiesHTTPProxy as String] = proxyHost
            proxyDict[kCFNetworkProxiesHTTPPort as String] = proxyPort
        }
        
        config.connectionProxyDictionary = proxyDict
        
        print("[ProxyManager] Per-request proxy applied: \(proxyType)://\(proxyHost):\(proxyPort)")
    }
    
    // MARK: - Proxy Testing
    
    /**
     * Test proxy connection
     */
    public func testProxy(_ proxyConfig: JSObject, testUrl: String?, completion: @escaping (JSObject) -> Void) {
        let startTime = Date()
        
        let proxyType = proxyConfig["type"] as? String ?? "http"
        let proxyHost = proxyConfig["host"] as? String ?? ""
        let proxyPort = proxyConfig["port"] as? Int ?? 8080
        
        guard !proxyHost.isEmpty else {
            completion([
                "success": false,
                "error": "Proxy host is required"
            ])
            return
        }
        
        let urlString = testUrl ?? "https://www.google.com"
        guard let url = URL(string: urlString) else {
            completion([
                "success": false,
                "error": "Invalid test URL"
            ])
            return
        }
        
        // Configure session with proxy
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 10
        config.timeoutIntervalForResource = 10
        
        var proxyDict: [AnyHashable: Any] = [:]
        
        switch proxyType.lowercased() {
        case "socks4", "socks5", "socks":
            proxyDict[kCFStreamPropertySOCKSProxyHost as String] = proxyHost
            proxyDict[kCFStreamPropertySOCKSProxyPort as String] = proxyPort
            
            if let username = proxyConfig["username"] as? String, !username.isEmpty {
                proxyDict[kCFStreamPropertySOCKSUser as String] = username
                proxyDict[kCFStreamPropertySOCKSPassword as String] = proxyConfig["password"] as? String ?? ""
            }
            
        case "https":
            proxyDict[kCFNetworkProxiesHTTPSEnable as String] = true
            proxyDict[kCFNetworkProxiesHTTPSProxy as String] = proxyHost
            proxyDict[kCFNetworkProxiesHTTPSPort as String] = proxyPort
            
        default:
            proxyDict[kCFNetworkProxiesHTTPEnable as String] = true
            proxyDict[kCFNetworkProxiesHTTPProxy as String] = proxyHost
            proxyDict[kCFNetworkProxiesHTTPPort as String] = proxyPort
        }
        
        config.connectionProxyDictionary = proxyDict
        
        let session = URLSession(configuration: config)
        
        var request = URLRequest(url: url)
        request.httpMethod = "HEAD"
        
        let task = session.dataTask(with: request) { _, response, error in
            let responseTime = Date().timeIntervalSince(startTime) * 1000
            
            if let error = error {
                completion([
                    "success": false,
                    "responseTime": Int(responseTime),
                    "error": error.localizedDescription
                ])
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse else {
                completion([
                    "success": false,
                    "responseTime": Int(responseTime),
                    "error": "Invalid response"
                ])
                return
            }
            
            let success = (200..<400).contains(httpResponse.statusCode)
            
            completion([
                "success": success,
                "responseTime": Int(responseTime),
                "statusCode": httpResponse.statusCode
            ])
            
            print("[ProxyManager] Proxy test completed: \(success ? "success" : "failed") in \(Int(responseTime))ms")
        }
        
        task.resume()
    }
    
    // MARK: - Statistics
    
    /**
     * Record successful request
     */
    public func recordSuccess() {
        requestCount += 1
        lastSuccessTime = Date()
        lastError = nil
    }
    
    /**
     * Record failed request
     */
    public func recordError(_ error: String) {
        requestCount += 1
        lastError = error
    }
}

File: android/src/main/java/com/capacitor/cors/ProxyManager.java
package com.capacitor.cors;

import android.util.Log;
import com.getcapacitor.JSObject;
import okhttp3.*;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Proxy;
import java.net.PasswordAuthentication;
import java.net.Authenticator;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Proxy Manager for handling network proxy configurations
 * Supports HTTP, HTTPS, SOCKS4, and SOCKS5 proxies
 */
public class ProxyManager {
    private static final String TAG = "ProxyManager";
    
    // Proxy configuration
    private boolean enabled = false;
    private String type = "http";
    private String host = "";
    private int port = 8080;
    private String username = null;
    private String password = null;
    private String[] bypass = new String[0];
    private boolean applyToAll = true;
    
    // Statistics
    private final AtomicInteger requestCount = new AtomicInteger(0);
    private final AtomicLong lastSuccessTime = new AtomicLong(0);
    private String lastError = null;
    
    public ProxyManager() {
    }
    
    /**
     * Configure global proxy settings
     */
    public void setConfig(JSObject config) {
        if (config == null) {
            clearConfig();
            return;
        }
        
        this.enabled = config.getBoolean("enabled", false);
        this.type = config.getString("type", "http");
        this.host = config.getString("host", "");
        this.port = config.getInteger("port", 8080);
        this.username = config.getString("username", null);
        this.password = config.getString("password", null);
        this.applyToAll = config.getBoolean("applyToAll", true);
        
        // Parse bypass list
        try {
            com.getcapacitor.JSArray bypassArray = config.getJSArray("bypass");
            if (bypassArray != null) {
                this.bypass = new String[bypassArray.length()];
                for (int i = 0; i < bypassArray.length(); i++) {
                    this.bypass[i] = bypassArray.getString(i);
                }
            }
        } catch (Exception e) {
            this.bypass = new String[0];
        }
        
        Log.i(TAG, "Proxy configured: " + type + "://" + host + ":" + port + " (enabled: " + enabled + ")");
    }
    
    /**
     * Clear proxy configuration
     */
    public void clearConfig() {
        this.enabled = false;
        this.type = "http";
        this.host = "";
        this.port = 8080;
        this.username = null;
        this.password = null;
        this.bypass = new String[0];
        this.applyToAll = true;
        this.lastError = null;
        
        Log.i(TAG, "Proxy configuration cleared");
    }
    
    /**
     * Get current proxy configuration as JSObject
     */
    public JSObject getConfig() {
        if (!enabled) {
            return null;
        }
        
        JSObject config = new JSObject();
        config.put("enabled", enabled);
        config.put("type", type);
        config.put("host", host);
        config.put("port", port);
        if (username != null) {
            config.put("username", username);
        }
        config.put("applyToAll", applyToAll);
        
        com.getcapacitor.JSArray bypassArray = new com.getcapacitor.JSArray();
        for (String b : bypass) {
            bypassArray.put(b);
        }
        config.put("bypass", bypassArray);
        
        return config;
    }
    
    /**
     * Get proxy status
     */
    public JSObject getStatus() {
        JSObject status = new JSObject();
        status.put("active", enabled && !host.isEmpty());
        status.put("requestCount", requestCount.get());
        
        if (lastError != null) {
            status.put("lastError", lastError);
        }
        
        long successTime = lastSuccessTime.get();
        if (successTime > 0) {
            status.put("lastSuccessTime", successTime);
        }
        
        if (enabled) {
            status.put("config", getConfig());
        }
        
        return status;
    }
    
    /**
     * Check if proxy is enabled and configured
     */
    public boolean isEnabled() {
        return enabled && !host.isEmpty();
    }
    
    /**
     * Check if URL should bypass proxy
     */
    public boolean shouldBypass(String url) {
        if (bypass == null || bypass.length == 0) {
            return false;
        }
        
        try {
            java.net.URL parsedUrl = new java.net.URL(url);
            String urlHost = parsedUrl.getHost();
            
            for (String pattern : bypass) {
                if (matchesPattern(urlHost, pattern)) {
                    return true;
                }
            }
        } catch (Exception e) {
            // Ignore parsing errors
        }
        
        return false;
    }
    
    /**
     * Simple wildcard pattern matching for bypass list
     */
    private boolean matchesPattern(String host, String pattern) {
        if (pattern.startsWith("*.")) {
            // Wildcard domain match
            String suffix = pattern.substring(1);
            return host.endsWith(suffix) || host.equals(pattern.substring(2));
        } else if (pattern.endsWith(".*")) {
            // Wildcard IP match
            String prefix = pattern.substring(0, pattern.length() - 1);
            return host.startsWith(prefix);
        } else {
            return host.equalsIgnoreCase(pattern);
        }
    }
    
    /**
     * Create OkHttp Proxy object
     */
    public Proxy createProxy() {
        if (!isEnabled()) {
            return Proxy.NO_PROXY;
        }
        
        Proxy.Type proxyType;
        switch (type.toLowerCase()) {
            case "socks4":
            case "socks5":
            case "socks":
                proxyType = Proxy.Type.SOCKS;
                break;
            case "http":
            case "https":
            default:
                proxyType = Proxy.Type.HTTP;
                break;
        }
        
        return new Proxy(proxyType, new InetSocketAddress(host, port));
    }
    
    /**
     * Create proxy authenticator for OkHttp
     */
    public Authenticator createAuthenticator() {
        if (!isEnabled() || username == null || username.isEmpty()) {
            return null;
        }
        
        return new Authenticator() {
            @Override
            public Request authenticate(Route route, Response response) throws IOException {
                if (response.request().header("Proxy-Authorization") != null) {
                    // Already attempted authentication
                    return null;
                }
                
                String credential = Credentials.basic(username, password != null ? password : "");
                return response.request().newBuilder()
                    .header("Proxy-Authorization", credential)
                    .build();
            }
        };
    }
    
    /**
     * Apply proxy configuration to OkHttpClient.Builder
     */
    public void applyToClientBuilder(OkHttpClient.Builder builder) {
        if (!isEnabled()) {
            return;
        }
        
        builder.proxy(createProxy());
        
        Authenticator authenticator = createAuthenticator();
        if (authenticator != null) {
            builder.proxyAuthenticator(authenticator);
        }
        
        Log.d(TAG, "Proxy applied to HTTP client: " + type + "://" + host + ":" + port);
    }
    
    /**
     * Apply proxy configuration to OkHttpClient.Builder for a specific request
     * Takes into account per-request proxy config and bypass list
     */
    public void applyToClientBuilder(OkHttpClient.Builder builder, String url, JSObject requestProxy) {
        // Check if request has its own proxy config
        if (requestProxy != null && requestProxy.getBoolean("enabled", false)) {
            applyRequestProxy(builder, requestProxy);
            return;
        }
        
        // Check if global proxy should apply
        if (!isEnabled() || !applyToAll) {
            return;
        }
        
        // Check bypass list
        if (shouldBypass(url)) {
            Log.d(TAG, "Bypassing proxy for URL: " + url);
            return;
        }
        
        applyToClientBuilder(builder);
    }
    
    /**
     * Apply per-request proxy configuration
     */
    private void applyRequestProxy(OkHttpClient.Builder builder, JSObject proxyConfig) {
        String proxyType = proxyConfig.getString("type", "http");
        String proxyHost = proxyConfig.getString("host", "");
        int proxyPort = proxyConfig.getInteger("port", 8080);
        String proxyUser = proxyConfig.getString("username", null);
        String proxyPass = proxyConfig.getString("password", null);
        
        if (proxyHost.isEmpty()) {
            return;
        }
        
        Proxy.Type type;
        switch (proxyType.toLowerCase()) {
            case "socks4":
            case "socks5":
            case "socks":
                type = Proxy.Type.SOCKS;
                break;
            default:
                type = Proxy.Type.HTTP;
                break;
        }
        
        builder.proxy(new Proxy(type, new InetSocketAddress(proxyHost, proxyPort)));
        
        if (proxyUser != null && !proxyUser.isEmpty()) {
            builder.proxyAuthenticator((route, response) -> {
                if (response.request().header("Proxy-Authorization") != null) {
                    return null;
                }
                String credential = Credentials.basic(proxyUser, proxyPass != null ? proxyPass : "");
                return response.request().newBuilder()
                    .header("Proxy-Authorization", credential)
                    .build();
            });
        }
        
        Log.d(TAG, "Per-request proxy applied: " + proxyType + "://" + proxyHost + ":" + proxyPort);
    }
    
    /**
     * Test proxy connection
     */
    public JSObject testProxy(JSObject proxyConfig, String testUrl) {
        JSObject result = new JSObject();
        long startTime = System.currentTimeMillis();
        
        try {
            String proxyType = proxyConfig.getString("type", "http");
            String proxyHost = proxyConfig.getString("host", "");
            int proxyPort = proxyConfig.getInteger("port", 8080);
            String proxyUser = proxyConfig.getString("username", null);
            String proxyPass = proxyConfig.getString("password", null);
            
            if (proxyHost.isEmpty()) {
                result.put("success", false);
                result.put("error", "Proxy host is required");
                return result;
            }
            
            // Default test URL
            if (testUrl == null || testUrl.isEmpty()) {
                testUrl = "https://www.google.com";
            }
            
            // Create proxy
            Proxy.Type type;
            switch (proxyType.toLowerCase()) {
                case "socks4":
                case "socks5":
                case "socks":
                    type = Proxy.Type.SOCKS;
                    break;
                default:
                    type = Proxy.Type.HTTP;
                    break;
            }
            
            Proxy proxy = new Proxy(type, new InetSocketAddress(proxyHost, proxyPort));
            
            // Build client
            OkHttpClient.Builder builder = new OkHttpClient.Builder()
                .proxy(proxy)
                .connectTimeout(10, TimeUnit.SECONDS)
                .readTimeout(10, TimeUnit.SECONDS);
            
            if (proxyUser != null && !proxyUser.isEmpty()) {
                final String user = proxyUser;
                final String pass = proxyPass;
                builder.proxyAuthenticator((route, response) -> {
                    if (response.request().header("Proxy-Authorization") != null) {
                        return null;
                    }
                    String credential = Credentials.basic(user, pass != null ? pass : "");
                    return response.request().newBuilder()
                        .header("Proxy-Authorization", credential)
                        .build();
                });
            }
            
            OkHttpClient client = builder.build();
            
            // Make test request
            Request request = new Request.Builder()
                .url(testUrl)
                .head()
                .build();
            
            Response response = client.newCall(request).execute();
            long responseTime = System.currentTimeMillis() - startTime;
            
            result.put("success", response.isSuccessful() || response.isRedirect());
            result.put("responseTime", responseTime);
            result.put("statusCode", response.code());
            
            response.close();
            
            Log.i(TAG, "Proxy test successful: " + responseTime + "ms");
            
        } catch (Exception e) {
            long responseTime = System.currentTimeMillis() - startTime;
            result.put("success", false);
            result.put("responseTime", responseTime);
            result.put("error", e.getMessage());
            
            Log.e(TAG, "Proxy test failed: " + e.getMessage());
        }
        
        return result;
    }
    
    /**
     * Record successful request
     */
    public void recordSuccess() {
        requestCount.incrementAndGet();
        lastSuccessTime.set(System.currentTimeMillis());
        lastError = null;
    }
    
    /**
     * Record failed request
     */
    public void recordError(String error) {
        requestCount.incrementAndGet();
        lastError = error;
    }
}
